Handle product tags in v2 page

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Integrate taggable component in product page v2, and integrate `SetProductTagsCommand` into the `SEOCommandsBuilder`
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | Go to product experimental page and edit some tags
| Possible impacts? | ~


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/27328)
<!-- Reviewable:end -->

### Breaking Change

- `SetProductTagsCommand` now throws a `PrestaShop\PrestaShop\Core\Exception\InvalidArgumentException` exception instead of a `RuntimeException` when empty tags is provided as argument
- `TagRepository` now needs two parameters to handle Doctrine connection queries
- `ProductTagUpdater` now depends on `ProductIndexationUpdater`

jolelievre committed:
Integrate tag component in product page SEO tab

jolelievre committed:
SEOCommandsBuilder also handles the SetProductTagsCommand

saulaski commented on src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php:
```diff
@@ -172,7 +174,11 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    '%s %s',
+                    $this->trans('To add tags, click in the field, write something, and then press the "Enter" key.', 'Admin.Shopparameters.Help'),
+                    $invalidCharsText
+                ),
```
Shouldn't we update the translation message to include the invalid characters as a placeholder ?

jolelievre replied:
We could but it would include a new wording that needs to be translated in all languages by the crowdin community As much as possible we try to use the existing wording when available
But I'll test your proposal and we'll ask @Julievrz her opinion about this

Julievrz replied:
I'm not against modifying an existing string if we need to be more specific (as long as it's not a patch version 😝). However, we should avoid giving essential information such as invalid characters in a placeholder. Because placeholders don't appear permanently on the page if users forget what was written they will have to delete everything to reveal the placeholder again. 

Actually, the wording had been modified in the mockups:

![image](https://user-images.githubusercontent.com/70583503/149972938-7df50ef2-2df2-4170-9a7b-b1efec35ea95.png)

Maybe we could add the information about invalid characters in the help text, next to "Separate tags with a comma." (And remove "Separate each tag with a comma" in the first sentence, as it is kind of repetitive.)

jolelievre replied:
I agree for the paceholder Except this text is not a placeholder it's a help message under the input so it should be fine. I updated a few things to match the mockup. I didn't see that the tag components works with "Enter" AND comma key so I updated the old wording a bit to add both.

This is how it looks like now:

![Capture d’écran 2022-01-25 à 17 08 40](https://user-images.githubusercontent.com/13801017/151014083-714f77c5-d21e-45cf-a746-23e9653f20ae.png)

Both texts are maybe a bit redundant now so I'm open for suggestions :)

sowbiba commented on src/Core/Form/IdentifiableObject/CommandBuilder/Product/SEOCommandsBuilder.php:
```diff
@@ -61,6 +63,18 @@ public function buildCommands(ProductId $productId, array $formData): array
         }
 
-        return [$command];
+        if (!empty($seoData['tags'])) {
```
```suggestion
        if (!empty($seoData['tags']) && is_array($seoData['tags'])) {
```

For me, nothing says it's an array

saulaski replied:
Just in case the tags are an array OR an iterable object:
```suggestion
        if (!empty($seoData['tags']) && is_iterable($seoData['tags'])) {
```

jolelievre replied:
But we don't usually test the type of array fields in all our builders, we kind of expect that the format from the form matches our expectation Not all fields are mandatory to be here, so we check their presence, but when they are present they should be in the right format

It's already the case for all the translatable fields for example, `$seoData['meta_title']` is an array of values indexed by language ID, we inject the value in the set method without pre-checking because if it's not it will break (because of strong typing) and we want it to break Unexpected format should result in an error not be ignored

But in this case, you're right since the value is interpreted and reformatted before being injected in the command constructor we should check it, but I'll throw an exception instead if the field is not an array

And I don't think `is_iterable` is necessary, all our data are flattened array so `is_array` is enough Using `is_iterable` would sort of imply that another type than array is usable It's not expected so I prefer to insist on the array

saulaski replied:
Your point makes sense... and I just can't trust arrays :D

Progi1984 replied:
I approve @saulaski

matks reviewed (commented):
Nothing to add to my peers review 👍

jolelievre committed:
Improve invalid chars wording

jolelievre committed:
Improve check of tags input

jolelievre commented:
@matks @saulaski @sowbiba feedback addressed :)

Julievrz reviewed (commented):
A comment about wording

jolelievre committed:
Udpate display and wording for tags component

jolelievre committed:
Improve tag spacing

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php:
```diff
@@ -172,7 +173,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'label' => $this->trans('Tags', 'Admin.Catalog.Feature'),
                 'label_tag_name' => 'h2',
-                'help' => $this->trans('Use a comma to create separate tags. E.g.: dress, cotton, party dresses.', 'Admin.Catalog.Help'),
+                'label_subtitle' => $this->trans('Enter the keywords that customers might search for when looking for this product. Separate each tag with a comma.', 'Admin.Catalog.Feature'),
```
```suggestion
                'label_subtitle' => $this->trans('Enter the keywords that customers might search for when looking for this product.', 'Admin.Catalog.Feature'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php:
```diff
@@ -172,7 +173,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'label_subtitle' => $this->trans('Enter the keywords that customers might search for when looking for this product. Separate each tag with a comma.', 'Admin.Catalog.Feature'),
+                'help' => sprintf(
+                    '%s %s',
+                    $this->trans('To add tags, click in the field, write something, and then press the "Enter" key or comma.', 'Admin.Shopparameters.Help'),
```
```suggestion
                    $this->trans(' Separate each tag with a comma or press the Enter key.', 'Admin.Shopparameters.Help'),
```

Wdyt?

jolelievre committed:
Improve wording

kpodemski reviewed (changes requested):
Rebase needed :-(

jolelievre commented:
Hi @matks @saulaski 
I need your re-approval on this one after rebase for conflicts

saulaski reviewed (dismissed):
LGTM and don't forget to rebase of course :)

khouloudbelguith reviewed (dismissed):
Hi @jolelievre,

I have some issues:

**First issue**
1. Enable Experimental product page
2. Go to BO > Catalog > Product V2
3. Add a new product
4. See an error: the scroll is displayed

![image](https://user-images.githubusercontent.com/16067358/154707550-374cb2e8-4b7e-47ed-b6b0-38a1be0cef2b.png)

The same for the file 
![image](https://user-images.githubusercontent.com/16067358/154709985-d8359aa1-0d1a-436f-a8d0-401e3a066a9d.png)

**Second issue**
1. Enable Experimental product page
2. Go to BO > Catalog > Product V2
3. Edit a product
4. In the SEO tab, add a tag and save
5. Go to FO > Search with the tag
6. the product is not displayed in the search results

PS: if add a new product => ok

I attached a screen record
https://user-images.githubusercontent.com/16067358/154710815-a412984e-c3dc-4b44-9185-b2984fb41b55.mp4

**Third issue**
![image](https://user-images.githubusercontent.com/16067358/154711390-1fb47356-fb37-46c4-bd51-8b8dd001efc1.png)

**Fourth issue**
I have a behavior, I don't know if it is a bug or not
But, at a special character, some characters are not displayed.

https://user-images.githubusercontent.com/16067358/154712618-0c15dce8-74fa-40bf-93db-d30825bd79d3.mp4


Thanks!

jolelievre committed:
Fix LocalizedTags setter when input is empty to ensure an array is always present instead of null

jolelievre committed:
Update product indexation when the product tags are updated and it is active

jolelievre committed:
Add behat test related to search in front office to ensure the updated tags are used for the search

jolelievre committed:
Handle min width for tag input field

jolelievre commented:
Hi @khouloudbelguith 

I fixed most of your feedback:

#### First issue
Could not reproduce it, it's not related to the PR scope and I don't see any modifications in this PR that could impact this modal So I would think it's a cache issue or something related to your environment but nothing to fix in this PR anyway

#### Second issue
I didn't think about it but when the tags are updated the indexation of the product must be updated as well, this should work fine now.

#### Third issue
Not relevant for this PR and it was already present in the legacy page I'm really wondering if testing such values is relevant in most cases Anyway it could be fixed in a dedicated issue but not in this PR

#### Fourth issue
Not exactly the scope of this PR but I searched a little for this one, it turns out there is some kind of hidden input that is used by the tag component which was a bit small I forced its minimum width to 768px just like in legacy page You will see the same behavior on large screen with very long words as well but IMHO 768px is already wide enough

Progi1984 commented on tests/Integration/Behaviour/Features/Scenario/Product/update_tags.feature:
```diff
@@ -44,25 +46,74 @@ Feature: Update product tags from Back Office (BO)
       | en-US  |       |
       | fr-FR  |       |
+
+  Scenario: Tags can be used for product search
```
:heart:

Progi1984 reviewed (commented):
A last feedback from @saulaski and it's approved.

jolelievre commented on src/Core/Form/IdentifiableObject/CommandBuilder/Product/SEOCommandsBuilder.php:
```diff
@@ -61,6 +63,20 @@ public function buildCommands(ProductId $productId, array $formData): array
-        return [$command];
+        if (!empty($seoData['tags'])) {
+            $parsedTags = [];
+            if (is_array($seoData['tags'])) {
```
@saulaski @Progi1984 this is already handled here which is more efficient because it allows detecting invalid values and throw an exception (see the builder tests for invalid case and more details)

saulaski replied:
Is it the exception message ?
`Empty array of product tags provided in %s. To remove all product tags use %s.`
If so the message doesn't sound really explicit. Here the main issue is the `tags` option type expected to be an array of string tags for each language.
IMHO failure should get detected earlier at parsing level. Easier to debug.
But I'm being picky here, so I approve :)

jolelievre replied:
Yeah, you're right! The exception thrown was not related to the use case I was intending to check It was only a side effect Besides it made me realize the `RemoveAllProductTagsCommand` was never handled by the builder (even though it actually works by setting a localized array with empty values..)

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php:
```diff
@@ -172,7 +173,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'label_subtitle' => $this->trans('Enter the keywords that customers might search for when looking for this product.', 'Admin.Catalog.Feature'),
+                'help' => sprintf(
+                    '%s %s',
+                    $this->trans('Separate each tag with a comma or press the Enter key.', 'Admin.Shopparameters.Help'),
```
```suggestion
                    $this->trans('Separate each tag with a comma or press the Enter key.', 'Admin.Catalog.Help'),
```

jolelievre committed:
Fix wording namespace

jolelievre committed:
Improve SEOCommandBuilder about empty tags use cases

jolelievre reviewed (commented):
@Progi1984 @saulaski @Julievrz I addressed your feedback it should be better now

saulaski commented on src/Core/Form/IdentifiableObject/CommandBuilder/Product/SEOCommandsBuilder.php:
```diff
@@ -44,23 +47,55 @@ public function buildCommands(ProductId $productId, array $formData): array
+        }
+
+        // Only return the command if some updates have been detected
+        if ($seoUpdate) {
```
Only a suggestion about naming and strict comparison, nothing mandatory :)
```suggestion
        if ($isSeoUpdated === true) {
```

saulaski replied:
This new behaviour might be complex to implement with the new multi-shop command builder. Any idea ?

jolelievre replied:
Actually, it won't, this is already handled internally by the command builder The commands are systematically instantiated and passed to the command builder

However, if it doesn't detect any fields in the data array the command is not returned and is "filtered" so it's the same principle as here actually, but easier 😉

jolelievre replied:
And I don't think the strict test is needed here, besides as you mentioned Your PR will replace this boolean via the CommandBuilder

Do you think I should refacto my PR to match yours and thus limit the conflicts later? Regardless of which one is merged first?

Julievrz reviewed (dismissed):
Wording ✅

Progi1984 commented:
@jolelievre Could you rebase ?

saulaski commented on src/Core/Form/IdentifiableObject/CommandBuilder/Product/SEOCommandsBuilder.php:
```diff
@@ -44,23 +47,55 @@ public function buildCommands(ProductId $productId, array $formData): array
+                    $parsedTags[$langId] = !empty($rawTags) ? explode(',', $rawTags) : [];
+                }
+
+                $commands[] = new SetProductTagsCommand(
+                    $productId->getValue(),
+                    $parsedTags
+                );
```
Question out of scope
This setter method expects 2 arguments, and the new multi-shop command builder can handle only 1 argument at the moment.
I'm working now on the SEO command builder, and I already had to handle the case with `setRedirectOption` setter. I propose to pass extra arguments as a 4th parameter in command fields:
```php
        $config = new CommandBuilderConfig($this->modifyAllNamePrefix);
        $config
            // skipped code
            ->addMultiShopField('[seo][redirect_option][type]', 'setRedirectOption', CommandField::TYPE_STRING, [
                (int) ($formData['seo']['redirect_option']['target']['id'] ?? 0),
            ])
        ;
```
I doubt this trick could handle SEO tags here. What do you think about a more flexible implementation with a callback function ?
```php
        $config
            ->addMultiShopCallbackField(function (array $data, UpdateProductSeoCommand $command) {
               // Update the command here
            })
        ;
```

jolelievre replied:
For your case about `setRedirectOption` I agree the command builder is lacking something I can think about two possibilities:
- Either a callback as you suggest, but then we wouldn't use the PropertyAccess component in this case I guess (no problem with that)
- Or an array of property path instead of just one which can be done by two ways:
  - a real array, like `addMultiShopField(['[seo][redirect_option][type]', '[seo][redirect_option][type]']), ...` 
  - a string to decompose `addMultiShopField('[seo][redirect_option][type],[seo][redirect_option][type]'), ...` 

In any case we shouldn't change the possibility to use a simple string so that we don't need to change all existing use cases which are basic and simple to use.

jolelievre replied:
I wrote this as I was thinking ^^ But to avoid a function with multiple types a new dedicated function is probably better, and I didn't realize at first that your use case is not only about two parameters but also about dynamic selection of a value. So your suggestion seems the more fitting and the more adaptable one!!

I think you can integrate it in your PR, you should also add its "sister" function `addCallbackField` for cases not related with multi shop.

saulaski replied:
@jolelievre
The "extra arguments" solution is already implemented in my current work in progress. I used a simple array, but you already know I don't like arrays :)))

I like a lot your idea of property pathes listed as a single string. Simple and intuitive to use. But it doesn't solve the type field definitons to provide as 3rd argument. I'm affraid this solution can't handle the most tricky cases, like the SEO tags here.
Using a callback function is pretty much like moving a piece of code from one class to another.

saulaski replied:
@jolelievre I have a scary idea that might fix everything... but it means refactoring the current multi-shop handling.

Automatization of commands building impose strong constraints. The tricky issues mostly come from the factorization: 1 single configuration to handle each shop constraint (single or all). Perhaps not the most open solution here.
Instead why not using a standard setter method to let the builder update the command with the shop constraint ?
It means all multi-shop commands should implement the same `MultiShopCommandInterface`:
- `getShopConstraint(): ShopConstraint`
- `setShopConstraint(ShopConstraint $shopConstraint): self`

This interface could also be used as a disciminator to detect multi-shop commands in commands builder.
Then only 1 command is enough to define, and the builder only has to set the shop constraint. Maybe clone the command first to keep it safe.
It means we could mostly keep the single shop code as it is today, and define a single command to be reconfigured by the builder.

I can try to design such solution in my current task in progress.

saulaski replied:
After a quick chat with @jolelievre, `MultiShopCommandInterface` doesn't fit the need.

jolelievre reviewed (commented):
Thanks for your feedback @saulaski 

I have a few questions for you

saulaski commented:
@jolelievre 
> Do you think I should refacto my PR to match yours and thus limit the conflicts later? Regardless of which one is merged first?

I doubt the conflicts could be too crazy in this specific case. Don't bother mate :)
Conflicts might even become minimal if I implement the `MultiShopCommandInterface` idea I mentioned here:
https://github.com/PrestaShop/PrestaShop/pull/27328#discussion_r823664317

jolelievre commented:
By the way @saulaski I realized we don't need to worry about tags for multishop since they are global to all shops anyway So one less thing to think about 😅

saulaski commented:
@jolelievre Great news ! I first assumed that tags were also shop specific :)

jolelievre commented:
Hi @matks @saulaski
I need your re-approval on this one after rebase for conflicts please

khouloudbelguith reviewed (dismissed):
Hi @jolelievre,

Could you please check the Conflicting files?

Thanks!

jolelievre committed:
Improvement when all tags are empty return remove command

Amoifr commented on src/Adapter/Product/Repository/TagRepository.php:
```diff
@@ -110,4 +131,33 @@ public function deleteTagsByLanguage(ProductId $productId, LanguageId $languageI
+     *
+     * @return array Localized tags for a product
+     */
+    public function getProductTags(ProductId $productId): array
```
getProductTags or getLocalizedTags ?

Amoifr replied:
or getLocalizedProductTags ? (I think this is better according the it's used after)

Amoifr commented on src/Adapter/Product/Update/ProductTagUpdater.php:
```diff
@@ -86,5 +100,55 @@ public function setProductTags(Product $product, array $localizedTagsList): void
+
+            $currentTags = $this->stringifyTags($localizedProductTags[$localizedTags->getLanguageId()->getValue()] ?? []);
+            $updateTags = $this->stringifyTags($localizedTags->getTags() ?? []);
+            if ($currentTags != $updateTags) {
```
!== instead != ?

Amoifr commented on tests/Integration/Behaviour/Features/Context/CurrencyFeatureContext.php:
```diff
@@ -128,7 +128,9 @@ public function setNoCurrentCurrency()
     {
         $this->checkCurrencyWithNameExists($currencyName);
-        $this->getCurrentCart()->id_currency = $this->currencies[$currencyName]->id;
+        if ($this->getCurrentCart()) {
```
```suggestion
        if ($this->getCurrentCart() !== null) {
```

jolelievre committed:
Miscelleaneous improvement from review

jolelievre reviewed (commented):
Thanks @Amoifr I handled all your feedback

NeOMakinG commented on admin-dev/themes/new-theme/scss/pages/product/product_page.scss:
```diff
@@ -1338,6 +1338,33 @@ $product-page-padding-bottom: 80px !default;
+    }
+
+    .token-input {
+      min-width: 768px;
```
This thing will break the mobile layout, please don't!!

You should try to avoid it somehow, maybe a `width: 100%; max-width: 768px;` would do the job, overwise you need to find another way to do this

jolelievre replied:
I can't, there's a *** rule from the initial component scss
https://github.com/PrestaShop/PrestaShop/blob/46510cc8cfaa1f995dff6bc92e485914f8e57d40/admin-dev/themes/new-theme/scss/components/_toggable_field.scss#L48

It overrides all the width value, including some in-style value that seems to be computed by JS. This is a pain...

NeOMakinG commented on admin-dev/themes/new-theme/js/pages/product/product-map.ts:
```diff
@@ -162,6 +162,7 @@ export default {
     defaultDescription: '.serp-default-description',
     watchedDescription: '.serp-watched-description',
     watchedMetaUrl: '.serp-watched-url:input',
+    tagFields: 'input.js-taggable-field',
```
```suggestion
    tagFields: '.js-taggable-field',
```

At the same time than fixing the other comment, remove the input element selector, so in the future in case it's not an input, it won't break, the JS class seems enough

jolelievre replied:
No I did this on purpose because `js-taggable-field` is added on both the div container and the input and here I need to specifically target the input I don't know why the class is duplicated but my guess is it's related to the translatable input and its default integration on the prestashop form theme

NeOMakinG replied:
Ok, not related to the PR, but added a todo

NeOMakinG commented on admin-dev/themes/new-theme/js/pages/product/edit/product-seo-manager.ts:
```diff
@@ -74,5 +75,14 @@ export default class ProductSEOManager {
+      options: {
+        createTokensOnBlur: true,
+        delimiter: ',',
+        minWidth: '768px',
```
Damn, did I miss this somewhere? We need to make sure this will work on mobile

jolelievre replied:
This was copied from the legacy code https://github.com/PrestaShop/PrestaShop/blob/26e53e18ccb2a670955bd4f0ed077216787bd1d7/admin-dev/themes/default/js/bundle/product/form.js#L2296

I need this or the js component `tokenfield` overrides it with in-lint styling add by JS And without this min width styling I get the bug from this comment (fourth issue video): https://github.com/PrestaShop/PrestaShop/pull/27328#pullrequestreview-887432596

It's an input field that is hidden inside the component's container:
![Capture d’écran 2022-04-04 à 16 23 15](https://user-images.githubusercontent.com/13801017/161565835-3d96349e-9c25-4d7b-babc-ab4cd277d05a.png)

I added a media query to avoid this min width to apply on screens lesser than 768px So obviously there are some bigs left in some sizes and some specific words' length but as far as I'm concerned I've done enough Any further improvement would need to be handled by someone more skilled in integration The purpose of the PR was to edit the tags That works anything more I don't care anymore

jolelievre replied:
Bootstrap component doesn't allow to prevent this auto computing and I couldn't find a better solution If you have one be my guest

NeOMakinG reviewed (changes requested):
Can I get some screenshots in order to show how it reacts on mobile? Seems broken from a review point of view

jolelievre committed:
Bug fix

jolelievre reviewed (commented):
I've struggled enough on this *** PR that keeps being blocked for edge cases, bootstrap components, form theme or whatever problems not related to the initial scope

So if anyone wants to make it perfect bugless be my guest, take the PR or open an issue But I'm tired working on this 3 months staled PR that was supposed to get rid in a matter of days

Commit:
fix: review feedbacks

HanaRebaiQA reviewed (commented):
Hello @jolelievre 

This PR is checked by @khouloudbelguith , @marwachelly and me.

**First issue :** 

- First case

1. Add a tag from Product V2 page 
2. Go to BO > Shop parameters > Search > Tags page
3. Delete the created tag 
4. Go to FO and search with this tag
5. See error > The product still displayed in the search result

- Second case

1. Add a tag from Product V2 page 
2. Save & try to search with it in the FO => ok
3. Go back to this product, remove the tag and save
4. Go to FO and search with this tag
5. See error > The product still displayed in the search result

https://user-images.githubusercontent.com/92912932/162204173-f1e5963d-dffa-4322-a91f-99d2aba30fbf.mp4

**Second issue :** 

1. Create a product V2
2. In the SEO tab > add a new tag with a specified language (eg:english) 
3. **Enable** this product and save
4. Go to BO > Shop parameters > Search > Tags page
5. Edit the last tag created 
6. Change in another languages (eg: Frensh) and Save
7. Come back to the product 
8. Click on Save
9. See error >> 

- If **php is 7.4** 

![image](https://user-images.githubusercontent.com/92912932/162206990-bad9925e-ff5f-416a-a832-938fad0bf010.png)

- if **php is 8.0**

![image](https://user-images.githubusercontent.com/92912932/162207241-0dcfd64d-a38e-40c5-89be-c0c59543d158.png)

https://user-images.githubusercontent.com/92912932/162208602-86b57cb8-03c4-42a2-9cae-8c4249940ea4.mp4



**Third issue :** 

1. Create a product V2
2. In the SEO tab > add a new tag with a specified language (eg: English) 
3. **Disable** this product and save
4. Go to BO > Shop parameters > Search > Tags page
5. Edit the last tag created 
6. Change in another languages (eg: Frensh) and Save
7. Come back to the product 
8. Check the tag with the updated language 
9. See error >> the tag is not displayed

https://user-images.githubusercontent.com/92912932/162208563-606d00dc-5e5a-497c-ab0e-06c9cc9e194e.mp4


**Fourth issue :** 

1. Create a product V2
2. In the SEO tab > add a new tag which contains **"** or **°** 
3. Save > no alert is displayed

In the product V1, there is an alert
![image](https://user-images.githubusercontent.com/92912932/162210862-257c5c44-c5ea-4bba-b356-b55354c60985.png)


**Fifth issue :** 

1. Create a product V2
2. In the SEO tab > add a new tag with long expression (eg: `ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt`)
3. Add another tag (eg: test)
4. Click on the first tag
5. See error >> an empty space is displayed

![image](https://user-images.githubusercontent.com/92912932/162211584-23a3e377-98ac-469b-b8d8-fb702fa0c844.png)

In the product V1, the tag is well displayed
![image](https://user-images.githubusercontent.com/92912932/162211746-6b8fc112-e820-491f-beff-d576e5860b11.png)

https://user-images.githubusercontent.com/92912932/162212190-b6ae21d2-42e5-4427-b7c8-45ae94f1da6f.mp4

** Sixth issue :** 

1. Create a product V2
2. In the SEO tab > add a new tag with long expression (eg: `ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt`)
3. Save
4. See error > The tag is truncated and is replaced by this expression `tttttttttttttttttttttttttttttttt` in Product page and in Search > Tags page
Product page
![image](https://user-images.githubusercontent.com/92912932/162212867-fcd3b27e-a951-409f-9558-983f18c68da6.png)
Tags page


Could you please check ?

Thanks!

jolelievre commented:
Hi @HanaRebaiQA, @khouloudbelguith and @marwachelly 

I looked into your feedback and noticed this:

### First issue
- First case: remove from Tags list
This bug is more related to the Tag page which doesn't update correctly the relationship with the product and the search indexes

- Second case: remove product's tag
Related to product page, I need to update the search indexes when the tags are changed, this should be fixed in this PR (the bug is not present in old page)

### Second and third issue
Here also the problem comes from the Tag page because it doesn't correctly update the products associated to the edited tag. I agree that the old page doesn't have this issue but it's because it ignores the inconsistent state. The page v2 doesn't accept such inconsistency but the bug should be resolved at its core in the Tags page.

By the way third issue happens on both product pages. Since both issues can be fixed with the same bug fix I think they should be handled independently from this PR in a dedicated issue (like the first issue here by the way).

### Fourth issue

I agree there is a difference, but there is no such limitation on tags so I don't see why legacy page blocks this. Even if you create a tag manually via the Tags page it is allowed so to me the problem was on the old behaviour. However when you try a truly invalid character like `<` it is correctly blocked in Tags page, legacy product page and new product page, this seems more logical.

### Fifth issue

UX CSS bug to fix here

### Sixth issue
About the tag string length actually it's logical the tag length is limited to 32 characters in the DB so when longer strings are used they are truncated. You can see the same behaviour on old page. So this is related to the Tag feature itself not product page, it should be handled in a whole new feature issue if we want to change this. Among the questions: should we increase the length? should we limit the length in the tag component? or instead display an error explaining the value is too long? 
But all this is a new feature anyway and not related to this PR.

I'm gonna see with @MatShir to arbitrate which of these points should:
- be fixed in this PR
- be handled in another bug issue
- be handled in a feature issue

jolelievre committed:
Update product indexation when its tags are all deleted

jolelievre committed:
Add option to TaggableField to limit the number of characters

jolelievre committed:
Limit product tags length and remove CSS to have a long input

jolelievre commented:
Hi @HanaRebaiQA, @khouloudbelguith and @marwachelly

so I discussed with @MatShir last Friday and we agreed on the scope of these bugs and which ones should be fixed now or in a dedicated PR, sadly I believe he didn't have time to confirm this via a message. So I'll discuss this with @Robin-Fischer-PS maybe or we'll have to wait for next eek for confirmation. Anyways we can still focus on what has been fixed and validate if the behaviour is now ok, so here's a resume of what I did.

#### First issue, second case
The removal of tags **from product page** now reset the search indexes so the product no longer appears in search in FO.

#### Fifth issue and sixth issue
I originally was only supposed to fix the fifth issue regarding the UX bug, but it's actually also related to the sixth issue. A tag can only be 32 characters long so there's no reason for complicated use cases with long values or to handle long enough inputs in this case. So I cleaned the CSS a little but mainly I added an option on the tag component, so it is now limited to words (or sentences) which have max 32 characters.

### About the rest
Fourth issue is not a bug, the new behaviour is more consistent than the previous page. The other bugs are related to the Tags page and need to be fixed in a dedicate issue that **must** in 8.0.

If it's all ok for you we can create the issue for Tags page and maybe this PR would be closable after a last round of tests? (if no other bugs are found)

marwachelly reviewed (commented):
Hello @jolelievre,

I tested the PR and I noticed that the first issue, second case is well fixed from product page as you can see in the attached screen record :

https://user-images.githubusercontent.com/89604350/162771302-1709da70-a481-4f6d-9d41-e4aef9b8b4ea.mp4

But regarding the fifth issue and sixth issue I detected the following problems :
**First issue:**
 When adding a duplicated tag to the product with 32 characters (for exemple "Ensembles_des_produits_de_beauté") : the duplication is not deleted and the second tag is replaced by this "Ensembles_des_produits_de_beaut"
 

https://user-images.githubusercontent.com/89604350/162773044-d45ea2d0-3479-4735-a1f9-4f233e8e8fe3.mp4

**Second issue:**
When adding the tag "Ensembles_des_produits_de_beauté" as a second tag and save, the last character is deleted

https://user-images.githubusercontent.com/89604350/162774023-7c71dcf5-a21c-4862-9939-05739ec521fc.mp4

Could you please check?

Thanks,

jolelievre committed:
Fix bug for tag content with not trimmed spaces

jolelievre committed:
Fix builder tests after rebase

jolelievre commented:
Hi @marwachelly 

thanks for the feedback! I fixed both issues (it was actually the same bug). Just waiting for re-approval and we should be good to go

MeKeyCool commented on src/Adapter/Product/Repository/TagRepository.php:
```diff
@@ -110,4 +131,33 @@ public function deleteTagsByLanguage(ProductId $productId, LanguageId $languageI
+
+        $result = $qb->execute()->fetchAllAssociative();
+        if (empty($result)) {
+            return [];
```
If `$result` is empty, won't
```php
        $localizedTags = [];
        foreach ($result as $row) {
            $localizedTags[(int) $row['id_lang']][] = $row['name'];
        }

        return $localizedTags;
```
return `[]` ?

MeKeyCool commented on src/Adapter/Product/Update/ProductTagUpdater.php:
```diff
@@ -67,24 +75,78 @@ public function setProductTags(Product $product, array $localizedTagsList): void
+            $updateTags = $this->stringifyTags($localizedTags->getTags());
+            if ($currentTags !== $updateTags) {
+                return true;
+            }
```
Why creating a new function ?

Maybe I miss something but I would search something like
```php
$currentTags = $localizedProductTags[$localizedTags->getLanguageId()->getValue()] ?? [];
$updateTags = $localizedTags->getTags();
if (!empty(array_diff($currentTags, $updateTags))) {
  return true;
}
```

MeKeyCool commented on src/Core/Domain/Product/Command/SetProductTagsCommand.php:
```diff
@@ -77,12 +77,12 @@ public function getLocalizedTagsList(): array
     {
         if (empty($localizedTags)) {
-            throw new RuntimeException(sprintf(
+            throw new InvalidArgumentException(sprintf(
```
:heart:

MeKeyCool commented on src/Core/Domain/Product/ValueObject/LocalizedTags.php:
```diff
@@ -91,7 +91,9 @@ public function getTags(): array
+        foreach ($tags as $tag) {
             //skip empty value
             if (empty($tag)) {
                 continue;
```
Typically an action I would enjoy to log (would be useful to know that we try to set empty values).

MeKeyCool commented on src/Core/Form/IdentifiableObject/CommandBuilder/Product/SEOCommandsBuilder.php:
```diff
@@ -75,11 +78,39 @@ public function buildCommands(
+
+        $seoData = $formData['seo'] ?? [];
+        if (isset($seoData['tags'])) {
+            if (!empty($seoData['tags'])) {
```
Why chaining `isset(...)` and `empty(...)` tests ? Are you looking for special combinations ? Should I search detailed difference between isset and empty to understand your case ?

https://www.php.net/manual/fr/types.comparisons.php
I don't get the difference cause there are about type (string, int, ...) and you throw an exception if it is not an array.

Is it a way to apply the `else` condition to `$seoData['tags']='0'` or `$seoData['tags']=0` or `$seoData['tags']=[]` or `$seoData['tags']=false` ?

Maybe some comments would help to understand expected the behavior here :thinking: .

MeKeyCool commented on src/Core/Form/IdentifiableObject/CommandBuilder/Product/SEOCommandsBuilder.php:
```diff
@@ -75,11 +78,39 @@ public function buildCommands(
+                    );
+                }
+            } else {
+                $commands[] = new RemoveAllProductTagsCommand($productId->getValue());
```
I think it is useless.

If `$seoData['tags']` is empty, `foreach ($seoData['tags']` shouldn't be called and you should apply 
```php
if ($allEmpty) {
  $commands[] = new RemoveAllProductTagsCommand($productId->getValue());
}
```

MeKeyCool commented on tests/Integration/Behaviour/Features/Context/Domain/Product/SearchProductFeatureContext.php:
```diff
@@ -123,4 +126,42 @@ public function assertNoProductsFound(string $localeReference, string $search):
+            $productId = $this->getSharedStorage()->get($expectedProduct['product_id']);
+            $foundProduct = false;
+            foreach ($search['result'] as $product) {
+                if ((int) $product['id_product'] === $productId) {
```
Looks strange to make a cast with a strict type test.

marwachelly reviewed (commented):
Hello @jolelievre ,

Thank you for the fix. The issues mentioned in my last comment https://github.com/PrestaShop/PrestaShop/pull/27328#pullrequestreview-938196455 are well fixed. 

https://user-images.githubusercontent.com/89604350/163796665-6f87b866-7c87-4cb6-95f7-b06f6c09400b.mp4


But i noticed the following problems:

First issue:

There is still a problem for invalid characters when the shop have only one language. The alerte doesn't exist. 

https://user-images.githubusercontent.com/89604350/163796326-e9a62033-ef6d-405a-bb80-6cf5d22d77c0.mp4



Second issue:

When adding more than tag and Click on the first tag the empty space is still displayed.



https://user-images.githubusercontent.com/89604350/163797233-5a6fead2-63f6-48f1-84eb-8785e29d530c.mp4




Third issue:
1- add the languages : български език (Bulgarian) and Azərbaycan dili (Azerbaijani)
2- add the tag "MƏŞHUR MƏHSULLAR" ( contains characters from Azerbaijani language) when english language is selected
3- search the tag from FO when български език (Bulgarian) or Azərbaycan dili (Azerbaijani) is selected => The product is find with the tag
4- delete the tag from product page => the tag is deleted and the product is still found whith the tag

https://user-images.githubusercontent.com/89604350/163796072-174e0e89-ed05-4766-a40f-2f948617d4b7.mp4


The issue doesn't exist with 1.7.8.5.

https://user-images.githubusercontent.com/89604350/163796001-e1eed2ec-617f-4f89-93be-82ff4d517e32.mp4

Could you please check?

Thanks a lot,

jolelievre committed:
Display children errors in TranslatableType even when only one locale is present

jolelievre commented:
Hi @marwachelly 

thanks for the feedback. Here are my answers:

#### First issue
This bug is not related to the Tags feature, it comes from the `TranslatableType` component which is used everywhere, I think you will find the same bug in any translatable field when only one language is used, the children errors were not correctly displayed. Even though it's not related to this PR I still fixed it, however this could have impact on any translatable field in the BO now. But it seems to work fine. I don't want this PR to be blocked for weeks because of this, so if just fixing this bug is enough we can fix it here, if you consider it needs a heavy regression search all around the BO then we open a dedicated issue for this bug with appropriate QA everywhere.
Let me know which strategy you think is more fitting here?

#### Second issue
As explained before and discussed with @Robin-Fischer-PS and @NeOMakinG it is **impossible** to fix all the UX bugs from this PR, what we can hope at best is something usable for most cases. I can't fix it all in this PR because to fix everything correctly we need to fix the whole tag component everywhere in the BO and this needs to be handled in a dedicated issue and PR not here. So for now I think the current state is good enough.

#### Third issue
I couldn't reproduce this bug, and I'm not sure I understand it all. In the videos you associate the tags from different pages and different versions so IMHO it can't be considered as a regression, the scenarios are different in both cases. Seems like some steps are missing, or maybe the bug came from another previous manipulation that is not shown on the video?

I'm not saying there is no bug here, but it's out of scope of the initial PR here to integrate the tags component. This should be handled in a dedicated issue for this big which seems like an edge case or we will never be able to merge this PR in time for feature freeze.

Robin-Fischer-PS commented:
Hi there !

Since the feature freeze is close now, I think we should try to merge the PR ASAP.

In order to do so, we need : 

- A quick QA validation that the first issue is fixed (just the described bug, no exploratory session, because as @jolelievre said, it's not fixed everywhere)
- Creation of the two other issues, so that it can be fixed during the freeze
- Then merge the PR !

@marwachelly @khouloudbelguith @HanaRebaiQA : Could you test that the first issue is fixed, and create an issue for the third issue found by Marwa ? (Jo is taking care of creating the second issue)

Thanks all for your work !

jolelievre commented:
Hi @marwachelly 

here is the issue created regarding the UX bugs remaining on the tags that need to be fixed globally. https://github.com/PrestaShop/PrestaShop/issues/28288

I copied one bug that I know is remaining but maybe there are others, feel free to enrich the issue with missing bugs (only those related to UX of the component)

marwachelly reviewed (approved):
Hello @jolelievre,

I tested the PR, the alerte is well added when trying to save with invalid character and when the shop has only one language

https://user-images.githubusercontent.com/89604350/164022753-0a78dc99-e5a4-4e79-b4e7-6d87119d8c22.mp4

Thanks for the issue created. I will create the remaining ones. PR is QA :heavy_check_mark: 

Thank you for your work,

marwachelly commented:
Hello @jolelievre,

Here is the list of issues created:
https://github.com/PrestaShop/PrestaShop/issues/28296
https://github.com/PrestaShop/PrestaShop/issues/28295
https://github.com/PrestaShop/PrestaShop/issues/28294
https://github.com/PrestaShop/PrestaShop/issues/28293
https://github.com/PrestaShop/PrestaShop/issues/28291

Thanks,

jolelievre commented:
Thank you @marwachelly 

@MatShir maybe it's worth creating an epic to group all these issues? including this one https://github.com/PrestaShop/PrestaShop/issues/28288
